data/code/lib.c

```diff
@@ -589,3 +593,715 @@ void check_lap_ex()
 #endif
 }
 #endif
+
+// The importance of this remaining in its own section is that its functionality will be needed for
+// any future custom terrain types added, within the Extended Terrain Types patch or otherwise.
+#if EXTENDED_TERRAIN_TYPES
+// Check against all of the custom material flags enabled by the patch.
+int is_extended_terrain_type(const char terrain_type)
+{
+    return terrain_type == 0xB0;
+}
+
+
+int should_return_fake_code(char* const ground)
+{
+    int* ground_materials = (int*)(ground + 0x20);  // Pointer to the triangle's material in memory.
+    int ret = 0;
+
+    if (*ground_materials != 0)
+    {
+        const char terrain_type = *(char*)(*ground_materials + 0x16);  // Pointer to the collision flag.
+        ret = is_extended_terrain_type(terrain_type);
+    }
+
+    return ret;
+}
+
+// Game will search for a Splash object due to the material hash being used.
+// This nullifies that behaviour.
+void get_splash_code_inline()
+{
+    register char* const ground asm("r3");  // R3 is a CrsGround object.
+    asm("mr %r5, %r3"); // Stop registers from misbehaving; CrsGround stored for later.
```

Comment:
While this will likely work at the moment, the calling convention does not guarantee that `r5` would not be modified inside `should_return_fake_code()`. Have a look at the [Register usage and conventions](https://www.ibm.com/docs/en/aix/7.1?topic=overview-register-usage-conventions) page.

How about something like this:

```cpp
int get_splash_code_inline()
{
    register const char* const r3 asm("r3");  // R3 is a CrsGround object.

    const char* const ground = r3;
    if (should_return_fake_code(ground))
    {
        return 0;
    }

    return (int)ground;
}
```
Note that the signature of the function has changed to return `int` (as opposed to synthetically populating `r3`).

DISCLAIMER: Not tested.

Reply:
> return (int)ground;

I have never been able to get a hook function to `return` anything at all into `r3`.
The natural assumption would be that it should work, but as the current implementation of `get_splash_code_inline` with the dirty write into `r5` would imply; I initially wanted to `return` the correct value, and what exists is a mere compromise to make it work.

Of course, functions inserted and called by the hook function can `return` without any issue, but I had assumed that the inability to do so out of the main hook call was impossible by design/limitation.

I'm almost certainly doing something wrong, but I'm unsure exactly what. 
Attached below: I load 9 into `r3` and then attempt to `return` another number, which should override `r3` to my understanding. At the point where Dolphin warns of an invalid read, you can see `r3` is equal to 9.

![image](https://github.com/cristian64/mkdd-extender/assets/61329703/543e2cc8-5a12-4265-9e6a-75c7e61f6ca6)

Is there a reason why this wouldn't work that I'm not understanding? I thought, "it should!" - but no matter what I try, never.

Reply:
In that image you shared, the return type is `void`. Could that explain it? 

Also, if you enable the **Debug Output** option in the Extender, you should be able to see the assembly for the symbol, so you can compare how it looks like (`r3` in particular) without having to boot up the game.

Reply:
What a huge blunder. That's the worst thing I've done in a long time! Oh well, yes that was exactly the issue.

Passing `ground` casted to int would be wrong, but passing the right value isn't so bad. Now, I have a lot to think about...